content/articles/How-to-start-up-with-Socket-Programming-with-Java/index.md

```diff
@@ -0,0 +1,331 @@
+  
+public class Server {  
+  
+    public static void main(String[] args) throws IOException {   
+  Socket socket ;  
+  InputStreamReader inputStreamReader ;  
+  OutputStreamWriter outputStreamWriter ;  
+  BufferedReader bufferedReader ;  
+  BufferedWriter bufferedWriter ;  
+  ServerSocket serversocket ;  
+  
+serversocket = new ServerSocket(5000);  
+  
+ while (true) {  
+   try {  
+         
+         socket = serversocket.accept();  
+  
+inputStreamReader = new InputStreamReader(socket.getInputStream());  
+outputStreamWriter = new OutputStreamWriter(socket.getOutputStream());  
+
+bufferedReader = new BufferedReader(inputStreamReader);  
+bufferedWriter = new BufferedWriter(outputStreamWriter);  
+
+while (true)
+{  
+
+String msgFromClient = bufferedReader.readLine();  
+
+System.out.println("Client: " + msgFromClient);   
+
+bufferedWriter.write(" MSG Received"); 
+bufferedWriter.newLine();  
+bufferedWriter.flush();  
+
+if (msgFromClient.equalsIgnoreCase("BYE"))  
+  break;  
+  }  
+
+        socket.close();  
+        inputStreamReader.close();  
+        outputStreamWriter.close();  
+        bufferedReader.close();  
+        bufferedWriter.close();  
+  
+          } catch (IOException e) {  
+            e.printStackTrace();  
+          }  
+        }  
+    }  
+}
+```
+
+The output of this  program is:
+
+```
+Connected
+Client: Hey there 
+Client: am doing this for the first time
+Client: BYE
+Closing connection
+```
+
+### Client side Programming
+# Establish a Socket Connection
+This is an instance of a socket class and can be obtained in two ways: 
+    
+1. On the server side it can be used as a return value of the accept ( ) method.
+2. The builder of socket objects accepts the hostname of the server on which the server operates and the server's port number.
+
+```java
+socket = new Socket("localhost", 5000);
+```
+
+localhost - is a domain name like google.com, localhost takes you to your computer. It is resolved to IP address 127.0.0.1 .
+port - Port – This is built to connect with the server via the port number socket object.
+Two computers communicate through sockets.
+
+Client and Server communicate by writing to an output stream and reading from an input stream on the socket.
+
+```java
+inputStreamReader = new InputStreamReader(socket.getInputStream());  
+outputStreamWriter = new OutputStreamWriter(socket.getOutputStream());
+``` 
+
+The Scanner class is used to get input from the user from the console/terminal. System. in is a keyboard input connected input stream system.
+
+```java
+Scanner scanner = new Scanner(System.in);
+```
+
+### Methods used in Client Socket
+1. `public InputStream( getInputStream())` - returns an InputStream object for receiving data. Our Server will keep on receiving messages until client sends "BYE".
+2. `public OutputStream( getOutputStream())`- returns an OutputStream for sending data.
+3. `BufferReader( )` - Wraps inputStreamReader to improve efficiency
```

Comment:
some of the brackets have spaces in-between, others don't. Try to keep it consistent.